### PR TITLE
Spark 3.1 read support

### PR DESCRIPTION
**Pull request by davidrabinowitz**

### CODE REVIEW

File: spark-bigquery-dsv2/spark-3.1-bigquery/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryBatch.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * limitations under the License.
+ */
+package com.google.cloud.spark.bigquery.v2;
+
+import com.google.cloud.spark.bigquery.v2.context.BigQueryDataSourceReaderContext;
+import org.apache.spark.sql.connector.read.Batch;
+import org.apache.spark.sql.connector.read.InputPartition;
+import org.apache.spark.sql.connector.read.PartitionReaderFactory;
+
+public class BigQueryBatch implements Batch {
```
Comment (VinayLondhe14) [inline spark-bigquery-dsv2/spark-3.1-bigquery/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryBatch.java:23]: Is Batch support only relevant in Spark 3.1 or has the API changed? Don't see it for 2.4

File: spark-bigquery-dsv2/spark-3.1-bigquery/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryScanBuilder.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import org.apache.spark.sql.connector.read.SupportsReportStatistics;
+import org.apache.spark.sql.sources.Filter;
+import org.apache.spark.sql.types.StructType;
+
+/**
+ * Both Scan and ScanBuilder implementation, otherwise estimateStatistics() is not called due to bug
+ * in DataSourceV2Relation
+ */
+public class BigQueryScanBuilder
+    implements Scan,
```
Comment (VinayLondhe14) [inline spark-bigquery-dsv2/spark-3.1-bigquery/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryScanBuilder.java:34]: Is this a new API for Spark 3.1

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/SparkBigQueryConfig.java [hunk 3]
```diff
@@ -339,6 +353,14 @@ public static SparkBigQueryConfig from(
     return config;
   }
 
+  private static ImmutableMap<String, String> toLowerCaseKeysMap(Map<String, String> map) {
+    Map<String, String> result = new HashMap<>();
+    for (Map.Entry<String, String> enrty : map.entrySet()) {
```
[the remark this answers is not in the document]
Comment (davidrabinowitz) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/SparkBigQueryConfig.java:358]: Fixed

File: spark-bigquery-dsv2/spark-3.1-bigquery/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryBatch.java (hunk 1 above)
Comment (davidrabinowitz) [reply spark-bigquery-dsv2/spark-3.1-bigquery/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryBatch.java:23]: Yes, for a good overview please read [this post](http://blog.madhukaraphatak.com/spark-3-datasource-v2-part-2/).

File: spark-bigquery-dsv2/spark-3.1-bigquery/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryScanBuilder.java (hunk 2 above)
Comment (davidrabinowitz) [reply spark-bigquery-dsv2/spark-3.1-bigquery/src/main/java/com/google/cloud/spark/bigquery/v2/BigQueryScanBuilder.java:34]: Yes, for a good overview please read [this post](http://blog.madhukaraphatak.com/spark-3-datasource-v2-part-2/).

**Review by VinayLondhe14 — APPROVED** [review]

**Pull request merged**